Deploy Fluent Bit

alexandre-allard opened the pull request:
**Component**: salt, build, log

**Context**: 

**Summary**:

**Acceptance criteria**: 


---

Closes: #2701

alexandre-allard committed:
charts/log: initial import of fluent-bit helm charts

  helm repo add loki https://grafana.github.io/loki/charts
  helm repo update
  helm fetch -d charts --untar loki/fluent-bit

Refs: #2701

alexandre-allard committed:
charts/log: add fluent-bit helm configuration values

this file is used to overwrite default values.

Refs: #2701

alexandre-allard committed:
build: fetch fluent-bit image during build

Refs: #2701

alexandre-allard committed:
salt: render fluent-bit chart into Salt formula

  ./charts/render.py fluent-bit --namespace metalk8s-logging \
    --remove-manifest ConfigMap fluent-bit \
    charts/fluent-bit.yaml charts/fluent-bit/ \
    > salt/metalk8s/addons/logging/fluent-bit/deployed/chart.sls

Refs: #2701

alexandre-allard committed:
build: add fluent-bit formulas to salt tree

Refs: #2701

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

alexandre-allard committed:
tests: add scenario for logging pipeline

spawn a container that logs a single line
then try to retrieve it from Loki API

Refs: #2701

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2701-deploy-fluent-bit-from-charts` and the
destination branch `development/2.6`.

Please resolve the conflict on **the feature branch** (`improvement/2701-deploy-fluent-bit-from-charts`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2701-deploy-fluent-bit-from-charts
 $ git merge origin/development/2.6
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2701-deploy-fluent-bit-from-charts
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

alexandre-allard committed:
charts: fix render.py script

there is a bug introduced by the latest
change on how we handle CSC in this script,
the script fails if no --service-config or
--remove-manifest options are provided.

alexandre-allard committed:
salt: add fluent-bit configmap and init

Refs: #2701

alexandre-allard committed:
salt: add fluent-bit deployed to logging addon

Refs: #2701

TeddyAndrieux commented on charts/fluent-bit.yaml:
```diff
@@ -0,0 +1,126 @@
+  servicePath: /api/prom/push
+  # user: user
+  # password: pass
+config:
+  port: 2020
+  tenantID: '""'
+  batchWait: 1
+  batchSize: 10240
+  loglevel: warn
+  lineFormat: json
+  k8sLoggingParser: "Off"
+  removeKeys:
+    - kubernetes
+    - stream
+  autoKubernetesLabels: false
+  labels: '{job="fluent-bit"}'
+  labelMap:
+    kubernetes:
+      namespace_name: namespace
+      labels:
+        app: app
+        release: release
+      host: node
+      container_name: container
+      pod_name: instance
+    stream: stream
```
Since you do not use the config from the chart maybe it makes sense to comment this section and add a comment to say that we do not use this one from the chart as we cannot provide multiple inputs

TeddyAndrieux commented on salt/metalk8s/addons/logging/fluent-bit/deployed/configmap.sls:
```diff
@@ -0,0 +1,104 @@
+                DB             /run/fluent-bit/flb_kube.db
+                Mem_Buf_Limit  5MB
+            [INPUT]
+                Name              systemd
+                Tag               host.*
+                DB                /run/fluent-bit/flb_journal.db
```
Minor: I know we do not really care, but look like indentation is not the same as the other block

alexandre-allard replied:
Right, but even the configuration generated from the helm chart is not consistent with indentation. :D
I'll change it.

TeddyAndrieux commented on salt/metalk8s/addons/logging/fluent-bit/deployed/configmap.sls:
```diff
@@ -0,0 +1,104 @@
+            [FILTER]
+                Name           modify
+                Match          host.*
+                Rename         HOSTNAME node
```
Not sure we should rename it to node because hostname may be different from node name :/
I would instead retrieve the node name of the current host here if it's doable (or add a hostname label on k8s logs but not sure it's easy to do as well) 
If it's not possible or to much work, then ignore this comment and let's keep hostname as node for the moment

TeddyAndrieux replied:
PS look like hostname is not really good, on my test env I got some logs with `localhost` and `localhost.localdomain` as node  (so likely HOSTNAME)
Looking on log it's seems these logs are not really relevant (maybe we should filter out logs with hostname `localhost` and `localhost.localdomain` )

Edit: Ok log with `localhost` and `localhost.localdomain` are only logs at the beginning before `systemd-hostnamed` update the hostname

alexandre-allard replied:
It seems to be a little bit tricky to achieve that.
I can't find an easy way to retrieve the real hostname from the kubernetes logs or to retrieve the nodename from journal logs.
I've tried to use `$HOSTNAME` env var, but it's set to the container hostname... not really useful.
Plus, the helm chart does not allow to set up env vars based on pod specs... otherwise we could have done:
```
[...]
spec:
  containers:
    - env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
      [...]
```
And use the env var in fluent bit configuration.
Maybe we should just have 2 different labels `host` and `node`... or we should not use this chart.

TeddyAndrieux replied:
Ok let's use `hostname` and `pod` for the moment, I updated the dashboard accordingly

TeddyAndrieux commented on tests/post/steps/test_logging.py:
```diff
@@ -56,6 +73,60 @@ def _check_loki_ready():
+        ).format(template_path, write_template.stderr)
+
+        pod_creation_ts = int(time.time())
+        manage_static_pod = host.run(
+            "salt-call state.single metalk8s.static_pod_managed "
+            'name=%s source=%s',
+            manifest_path,
+            template_path,
+        )
```
Why create a static pod and using salt ? Would be better to create the pod using kubectl (or python k8s) (with a label selector if you really want it to be on bootstrap node, but I think we do not really care here)

alexandre-allard replied:
To not rely on kubectl, that's the only point, but I don't find a good reason for that.
I'll change it.

TeddyAndrieux commented on tests/post/steps/files/logger-pod.yaml.tpl:
```diff
@@ -0,0 +1,20 @@
+  name: $name
+  namespace: metalk8s-logging
+spec:
+  tolerations:
+  - key: "node-role.kubernetes.io/bootstrap"
+    operator: "Equal"
+    effect: "NoSchedule"
+  - key: "node-role.kubernetes.io/infra"
+    operator: "Equal"
+    effect: "NoSchedule"
```
FYI: Since for the moment you use static pod you do not need tolerations

TeddyAndrieux commented on salt/metalk8s/addons/logging/fluent-bit/deployed/configmap.sls:
```diff
@@ -0,0 +1,104 @@
+                  "release": "release"
+                },
+                "namespace_name": "namespace",
+                "pod_name": "instance"
```
It's a bit weird to have `instance` label being the pod name, why not name it `pod` ?

alexandre-allard replied:
This was defined this way in the ConfigMap generated with the helm chart, I've changed the configuration as little as possible.
But you're right, it's clearer with `pod`, I'll change it.

TeddyAndrieux commented on tests/post/steps/test_logging.py:
```diff
@@ -56,6 +72,49 @@ def _check_loki_ready():
+        name=pod_name,
+        image=utils_image,
+    )
+    manifest_path = str(MANIFESTS_PATH / "{}.yaml".format(pod_name))
+
+    with host.sudo():
+        write_manifest = utils.write_string(host, manifest_path, manifest)
+        assert write_manifest.rc == 0, (
+            "Failed to write logger Pod manifest '{}': {}"
+        ).format(manifest_path, write_manifest.stderr)
+
+        pod_creation_ts = int(time.time())
+        result = host.run('kubectl apply -f %s', manifest_path)
```
Wow, it's not really good here, you put the file iin the `/etc/kubernetes/manifest/` directory and you kubectl apply, you should only do the kubectl apply otherwise I think here you will create 2 pods: 
- one static pod `logger-test-bootstrap` on the bootstrap node (because of a YAML in the `/etc/kubernetes/manifest` directory)
- one pod `logger-test` on a "random" node (because of the kubectl apply)

IMHO you should use the same approach as in `tests/post/steps/test_dns.py` it's would be simpler

alexandre-allard replied:
True, don't know what I had in mind, I may need some sleep. :D
Thanks for catching this.

TeddyAndrieux reviewed (commented):
Just a comment about tests

alexandre-allard committed:
salt: add a service for each loki replica

This is needed for Fluent Bit to be able to
send every logs to each Loki instance.

Refs: #2682

alexandre-allard committed:
salt: add 1 output block per loki replica

We write all logs to each Loki replicas,
so we need to define one Output block,
in Fluent Bit configuration, per loki instance.

Refs: #2701

TeddyAndrieux commented on salt/metalk8s/addons/logging/loki/deployed/services.sls:
```diff
@@ -0,0 +1,29 @@
+            - name: http
+              port: 3100
+              protocol: TCP
+              targetPort: http
```
Does it works ? The only defined port for Loki is `http-metrics`

alexandre-allard replied:
No, it does not work, I've tested with http-metrics and I made a mistake rewriting the sls.
We can even see that tests do not pass.

TeddyAndrieux commented on salt/metalk8s/addons/logging/fluent-bit/deployed/configmap.sls:
```diff
@@ -0,0 +1,109 @@
+            [Output]
+                Name           loki
+                Match          *
+                Url            http://loki-{{ index }}:3100/api/prom/push
```
According to documentation, this endpoint is deprecated in favor of `loki/api/v1/push` but the new endpoint does not use the same POST format, fluent-bit do not handle the new format ?

https://grafana.com/docs/loki/latest/api/#post-apiprompush

alexandre-allard replied:
As said in a previous comment, I've changed as less as possible the generated configmap, this was the endpoint defined by the helm chart.
I'll try with the new route and change it accordingly if it does work.

alexandre-allard replied:
Looks like it works.

TeddyAndrieux reviewed (approved):
LGTM

alexandre-allard commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
